mlflow/tracing/export/mlflow.py

```diff
@@ -70,24 +76,159 @@ def export(self, root_spans: Sequence[ReadableSpan]):
                 # an MLflow model evaluation context
                 self._display_handler.display_traces([trace])
 
-            # Log the trace to MLflow
             self._log_trace(trace)
 
     def _log_trace(self, trace: Trace):
+        """Log the trace to MLflow backend."""
+        upload_tag_task = Task(
+            handler=self._client._upload_trace_spans_as_tag,
+            args=(trace.info, trace.data),
+            error_msg="Failed to log trace spans as tag to MLflow backend.",
+        )
+
+        upload_trace_data_task = Task(
+            handler=self._client._upload_trace_data,
+            args=(trace.info, trace.data),
+            error_msg="Failed to log trace to MLflow backend.",
+        )
+
+        upload_ended_trace_info_task = Task(
+            handler=self._client._upload_ended_trace_info,
+            args=(trace.info,),
+            error_msg="Failed to log trace to MLflow backend.",
+        )
+
+        if MLFLOW_ENABLE_ASYNC_LOGGING.get():
+            self._async_queue.put(upload_tag_task)
+            self._async_queue.put(upload_trace_data_task)
+            self._async_queue.put(upload_ended_trace_info_task)
+        else:
+            upload_tag_task.handle()
+            upload_trace_data_task.handle()
+            upload_ended_trace_info_task.handle()
+
+
+class Task:
+    """A dataclass to represent a task to be processed by the async trace export queue."""
+
+    def __init__(self, handler: Callable, args: Sequence, error_msg: str = ""):
+        self._handler = handler
+        self._args = args
+        self._error_msg = error_msg
+
+    def handle(self):
+        """Handle the task. This method must not raise any exception."""
         try:
-            self._client._upload_trace_spans_as_tag(trace.info, trace.data)
+            self._handler(*self._args)
         except Exception as e:
-            _logger.debug(f"Failed to log trace spans as tag to MLflow backend: {e}", exc_info=True)
-
-        # TODO: Make this async
-        # The trace is already updated in processor.on_end method
-        # so we just log to backend store here
-        try:
-            self._client._upload_trace_data(trace.info, trace.data)
-            self._client._upload_ended_trace_info(trace.info)
-        except Exception as e:
-            # avoid silent failures
             _logger.warning(
-                f"Failed to log trace to MLflow backend: {e}",
+                f"{self._error_msg} Error: {e}. For full traceback, set logging level to debug.",
                 exc_info=_logger.isEnabledFor(logging.DEBUG),
             )
+
+
+class AsyncTraceExportQueue:
```

Comment:
Do you feel that we should test the functionality of this queue handler (particularly, validating that the flush functionality clears the queues, that repeated activation calls don't create multiple callback instances, and that exceptions thrown from backend REST requests for accessing data in the tracking service don't cause the handler threads to die)?

Reply:
Yea I think it's good idea to have those tests. I believe existing tests/warnings cover most of those bugs, but having direct tests for the queue functionality make failure debugging much easier.